pyodata/v2/model.py

```diff
@@ -376,6 +379,13 @@ def to_literal(self, value):
 
         return super(EdmDateTimeTypTraits, self).to_literal(value.isoformat())
 
+    def to_json(self, value):
+        if isinstance(value, str):
+            return value
+
+        # Converts datetime into timestamp in milliseconds
+        return f'/Date({int((value - datetime.datetime(1970, 1, 1)).total_seconds()) * 1000})/'
```

Comment:
datetime.timestamp(value) * 1000

Reply:
``` python
import datetime

# From datetime to timestamp
date = datetime.datetime(2017, 12, 24, 19, 0)

r1 = int(date.timestamp() * 1000)
print(r1)
# outputs: 1514138400000
# That is in GTM 2017/12/24 18:00 or in GTM+1 2017/12/24 19:00

r2 = int((date - datetime.datetime(1970, 1, 1)).total_seconds()) * 1000
print(r2)
# outputs: 1514142000000
# That is in GTM 2017/12/24 19:00 or in GTM+1 2017/12/24 20:00


# I deducted we have to send the datetime in GTM format, because 1) There is another type called datetimeoffset which is
# for passing datetime with timezones 2) Also passing it in different timezone would yield strange result.
# See the example below

# Timestamp 1514142000000 represents 2017/12/24 19:00 GTM. Let's see what happens if we repeatably parse this value to
# datetime and back to timestamp
t = 1514142000000


def foo(timestap):
    value = datetime.datetime(1970, 1, 1) + datetime.timedelta(milliseconds=int(timestap))
    return value.timestamp() * 1000


for i in range(0, 10):
    t = foo(t)

print(t)
# outputs: 1514106000000
# After around 10 iteration the timestamps differ for about 36000000 milliseconds

# Let's reset the value and trie different parsing method
t = 1514142000000


def foo2(timestap):
    value = datetime.datetime(1970, 1, 1) + datetime.timedelta(milliseconds=int(timestap))
    return int((value - datetime.datetime(1970, 1, 1)).total_seconds()) * 1000


for i in range(0, 10):
    t = foo2(t)

print(t)
# outputs: 1514138400000
# After around 10 iteration the timestamps does not differ.
# Therefore I think this is correct method to parse datetime as parser should yield original value if fed it's output

```

Reply:
I am sorry, why did you paste that code?

Reply:
You can replicate my results yourself as enclosed code is runnable.

Reply:
Ah, so it is not accurate.

Reply:
Yes, thus I suggest not to replace it with int(datetime.timestamp(value) * 1000). Do you agree?

Reply:
well, if we want to stick to seconds only, we can `int(datetime.timestamp(value)) * 1000`

Reply:
I am a bit confused. I still think that the problem is due to timestamp handling of the timezones and not the rounding of floats. From my understanding, the timestamp in python contains timezone information but the ODATA definition of DateTime does not include that. The user would have to rather use DateTimeOffset. (DateTime was deprecated in ODATA V4 and solely replaced by DateTimeOffset to avoid this kind of confusion)

Reply:
You are right: https://docs.python.org/3/library/datetime.html#datetime.datetime.timestamp

```
Note

There is no method to obtain the POSIX timestamp directly from a naive datetime instance representing UTC time. If your application uses this convention and your system timezone is not set to UTC, you can obtain the POSIX timestamp by supplying tzinfo=timezone.utc:

timestamp = dt.replace(tzinfo=timezone.utc).timestamp()

or by calculating the timestamp directly:

timestamp = (dt - datetime(1970, 1, 1)) / timedelta(seconds=1)
```

So, I think that your implementation suffers from the same bug:
1. the variable value can hold datetime of whatever timezone
2. datetime.datetime(1970, 1, 1) returns datetime in your current timezone

I propose to use this:
`int(value.replace(tzinfo=timezone.utc).timestamp() * 1000)`

Please, don't forget to add a comment with the reference to the code - I think this information belongs to code.

I adapted your code:

```python
#!/usr/bin/python3

import datetime

# From datetime to timestamp
date = datetime.datetime(2017, 12, 24, 19, 0)

r1 = int(date.timestamp() * 1000)
print('date.timestamp=', r1)
# outputs: 1514138400000
# That is in GTM 2017/12/24 18:00 or in GTM+1 2017/12/24 19:00

r2 = int((date - datetime.datetime(1970, 1, 1)).total_seconds()) * 1000
print('date - datetime(1970, 1, 1)=', r2)
# outputs: 1514142000000
# That is in GTM 2017/12/24 19:00 or in GTM+1 2017/12/24 20:00


# I deducted we have to send the datetime in GTM format, because 1) There is another type called datetimeoffset which is
# for passing datetime with timezones 2) Also passing it in different timezone would yield strange result.
# See the example below

# Timestamp 1514142000000 represents 2017/12/24 19:00 GTM. Let's see what happens if we repeatably parse this value to
# datetime and back to timestamp
t = 1514142000000


def foo(timestap):
    value = datetime.datetime(1970, 1, 1) + datetime.timedelta(milliseconds=int(timestap))
    return value.replace(tzinfo=datetime.timezone.utc).timestamp() * 1000


for i in range(0, 10):
    t = foo(t)

print('timestamp * 10=', t)
# outputs: 1514106000000
# After around 10 iteration the timestamps differ for about 36000000 milliseconds

# Let's reset the value and trie different parsing method
t = 1514142000000


def foo2(timestap):
    value = datetime.datetime(1970, 1, 1, tzinfo=datetime.timezone.utc) + datetime.timedelta(milliseconds=int(timestap))
    return int((value - datetime.datetime(1970, 1, 1, tzinfo=datetime.timezone(datetime.timedelta(hours=1)))).total_seconds()) * 1000


for i in range(0, 10):
    t = foo2(t)

print('subtraction * 10=', t)
# outputs: 1514138400000
# After around 10 iteration the timestamps does not differ.
# Therefore I think this is correct method to parse datetime as parser should yield original value if fed it's output
```
```
date.timestamp= 1514138400000
date - datetime(1970, 1, 1)= 1514142000000
timestamp * 10= 1514142000000.0
subtraction * 10= 1514178000000
```
BTW: In the first iteration, I didn't set tzinfo to the subraction base datetime and I got this error:

```Traceback (most recent call last):
  File "timestamp.py", line 50, in <module>
    t = foo2(t)
  File "timestamp.py", line 46, in foo2
    return int((value - datetime.datetime(1970, 1, 1)).total_seconds()) * 1000
TypeError: can't subtract offset-naive and offset-aware datetimes
```

Reply:
I also noticed is it not the same case with timezones for to_literal?

Reply:
Please, if you want to fix something else, create a solo commit.